### PR TITLE
Cast paths to string before saving to h5

### PR DESCRIPTION
**Pull request by mscheltienne**
Related to: https://github.com/h5io/h5io/issues/113
I had several failures saving a forward model where the `mri` field was provided as a path object.

```
Traceback (most recent call last):
  File "/mnt/Isilon/9003_CBT_HNP_MEEG/temp-mathieu/meg-eeg-fMRI/subject-03.py", line 237, in <module>
    fwd.save(directory / "Sources" / "forward" / f"sub_{subject}-pos_{pos}-fwd.h5")
  File "<decorator-gen-341>", line 12, in save
  File "/pyvenv/mne/lib/python3.10/site-packages/mne/forward/forward.py", line 176, in save
    write_forward_solution(fname, self, overwrite=overwrite)
  File "<decorator-gen-345>", line 12, in write_forward_solution
  File "/pyvenv/mne/lib/python3.10/site-packages/mne/forward/forward.py", line 903, in write_forward_solution
    _write_forward_hdf5(fname, fwd)
  File "/pyvenv/mne/lib/python3.10/site-packages/mne/forward/forward.py", line 911, in _write_forward_hdf5
    write_hdf5(fname, dict(fwd=fwd), overwrite=True)
  File "/pyvenv/mne/lib/python3.10/site-packages/h5io/_h5io.py", line 111, in write_hdf5
    _triage_write(title, data, fid, comp_kw, str(type(data)),
  File "/pyvenv/mne/lib/python3.10/site-packages/h5io/_h5io.py", line 148, in _triage_write
    _triage_write(
  File "/pyvenv/mne/lib/python3.10/site-packages/h5io/_h5io.py", line 148, in _triage_write
    _triage_write(
  File "/pyvenv/mne/lib/python3.10/site-packages/h5io/_h5io.py", line 155, in _triage_write
    _triage_write(
  File "/pyvenv/mne/lib/python3.10/site-packages/h5io/_h5io.py", line 148, in _triage_write
    _triage_write(
  File "/pyvenv/mne/lib/python3.10/site-packages/h5io/_h5io.py", line 235, in _triage_write
    raise TypeError(err_str)
TypeError: unsupported type <class 'pathlib.PosixPath'> (in <class 'dict'>["fwd"]["src"][0]["mri_file"])
```

### CODE REVIEW

**Comment by mscheltienne** [comment]
@larsoner Related to this saving issue, I also got this error once I fixed my paths:

```
TypeError: unsupported type <class 'scipy.sparse._csr.csr_array'> (in <class 'dict'>["fwd"]["src"][0]["interpolator"])
```

Do you know what we should cast the `csr_array` to?

**Comment by larsoner** [comment]
I think in this case actually the error is really in the `fwd` object. FIF has no notion of `Path` just `str` so really the `str` casting should happen there to be more correct. In other words, the `fwd` dict shouldn't ever have a `Path` in it.

> Related to this saving issue, I also got this error once I fixed my paths:

You maybe/probably need to update h5io, `csr_array` support was added somewhat recently (last 6 months maybe?)

**Comment by mscheltienne** [comment]
> I think in this case actually the error is really in the fwd object. FIF has no notion of Path just str so really the str casting should happen there to be more correct. In other words, the fwd dict shouldn't ever have a Path in it.

Agree, as the FIFF writer might be broken too in this case. I'll have a look. 
In any case, adding a cast to str for any paths object which makes it way to the `h5io` writer until proper support is added should still be safer.

> You maybe/probably need to update h5io, csr_array support was added somewhat recently (last 6 months maybe?)

Oh, my bad. Not use to running stuff on a remote computer not maintained..

File: mne/utils/check.py [hunk 1]
```diff
@@ -167,7 +167,30 @@ def _import_h5py():
...
+            if isinstance(key, Path):
+                keys2cast.append(key)
+        for key in keys2cast:
+            data[key.as_posix()] = data.pop(key)
+        return data
+
+    def write_hdf5(fname, data, *args, **kwargs):
+        """Write h5 and cast all paths to string in data."""
+        data = cast_path_to_str(data)
+        h5io.write_hdf5(fname, data, *args, **kwargs)
```
Comment (mscheltienne) [inline mne/utils/check.py:191]: Up-to-you to keep this safeguard or not; I think the change in `_check_mri` fixes the issue here and should be backported. Let's see if CIs are green.

**Comment by larsoner** [comment]
Thanks @mscheltienne !

**Pull request merged**